### PR TITLE
WIP: support keepdims in numpy mean

### PR DESCRIPTION
**Pull request by charlotte12l**
My current implementation, I haven't tested it and I'm not sure if my understanding is correct. I added four new functions in `numpy/mean.hpp `:

- `auto mean(E const &expr, types::none_type axis, dtype d, std::false_type keepdims)`: simply copy the codes in `auto mean(E const &expr, types::none_type axis, dtype d)`
- `auto mean(E const &expr, long axis, dtype d, std::false_type keepdims)`: simply copy the codes in `auto mean(E const &expr, types::none_type axis, dtype d)`
- `auto mean(E const &expr, types::none_type axis, dtype d, std::true_type keepdims)`: reshape the value array to support `keepdims`, but I'm not sure how can I specify the return type for this function.

```
>>> np.mean(np.array([[[1,2],[3,4]],[[5,6],[7,8]]]),  keepdims=False)
4.5
>>> >>> np.mean(np.array([[[1,2],[3,4]],[[5,6],[7,8]]]),  keepdims=True)
array([[[4.5]]])
>>> np.mean(np.array([[[1,2],[3,4]],[[5,6],[7,8]]]),  keepdims=False).reshape((1,1,1))
array([[[4.5]]])
```
- `auto mean(E const &expr, long axis, dtype d, std::true_type keepdims)`: expand the array `axis` dim to support `keepdims`.

```
>>> np.mean(np.array([[[1,2],[3,4]],[[5,6],[7,8]]]), axis=2, keepdims=True)
array([[[1.5],
        [3.5]],

       [[5.5],
        [7.5]]])
>>> np.expand_dims(np.mean(np.array([[[1,2],[3,4]],[[5,6],[7,8]]]), axis=2, keepdims=False), 2)
array([[[1.5],
        [3.5]],

       [[5.5],
        [7.5]]])
```

Some concerns: I found that `keepdims` is the 5th argument in `numpy.mean` while `out` is the fourth argument and we didn't support `out`.  Will it be a problem...? 

cc @serge-sans-paille @rgommers

### CODE REVIEW

**Comment by rgommers** [comment]
> Some concerns: I found that `keepdims` is the 5th argument in `numpy.mean` while `out` is the fourth argument and we didn't support `out`. Will it be a problem...?

I think `out` can be added in the Pythran signature - but it must be `None`. If the user passes anything but `None`, then raise an exception.

**Comment by charlotte12l** [comment]
@serge-sans-paille Hey Serge, do you have any suggestions on declaring the return type for `auto mean(E const &expr, types::none_type axis, dtype d, std::true_type keepdims)`? ... I think we need to calculated the `out_shape`  so we can ` -> decltype(numpy::functor::asarray{}(sum(expr) / 
 typename dtype::type(expr.flat_size())).reshape(out_shape))` ...

File: pythran/backend.py [hunk 1]
```diff
@@ -955,7 +955,10 @@ def visit_Call(self, node):
             return "{}({})".format(func, ", ".join(args))
 
     def visit_Constant(self, node):
-        if node.value is None:
+        if node in self.immediates and isinstance(node.value, bool):
+            ret = "std::integral_constant<%s, %s>{}" % (
+                PYTYPE_TO_CTYPE_TABLE[type(node.value)], node.value)
```
Comment (serge-sans-paille) [inline pythran/backend.py:960]: `node.value` will print `True` instead of `true`. You should use `str(node.value).lower()` instead

**Comment by serge-sans-paille** [comment]
> @serge-sans-paille Hey Serge, do you have any suggestions on declaring the return type for `auto mean(E const &expr, types::none_type axis, dtype d, std::true_type keepdims)`? ... I think we need to calculated the `out_shape` so we can ` -> decltype(numpy::functor::asarray{}(sum(expr) / typename dtype::type(expr.flat_size())).reshape(out_shape))` ...

We could speciailize the case where the `axis` is `None`, but a good start would be `types::ndarray<typename E::dtype, types::array<long, E::value>>`

**Comment by serge-sans-paille** [comment]
Extra note : you should add a few test case in `pythran/tests/test_numpy_func0.py` with the explicit `keepdims`

**Comment by charlotte12l** [comment]
Hi @serge-sans-paille, could you help look at why the checks always fail the tests if you got some time? I tried the following tests but they always failed with `no matching function for call to object of type`

```
    def test_mean6(self):
        self.run_test("def np_mean6(a): from numpy import mean ; from numpy import float64; return mean(a, 2, float64, None, False)", numpy.array([[[1, 2], [3, 4.]]]), np_mean6=[NDArray[float,:,:,:]])
```
or
```
def test_mean6(self):
        self.run_test("def np_mean6(a): from numpy import mean ; return mean(a, 2, keepdims=False)", numpy.array([[[1, 2], [3, 4.]]]), np_mean6=[NDArray[float,:,:,:]])
```

**Comment by serge-sans-paille** [comment]
@charlotte12l some fixes here: https://github.com/serge-sans-paille/pythran/tree/fix/add_keepdims
The remaining parts to fix are probably beyond your scope, I'll try to do that today.

**Comment by serge-sans-paille** [comment]
Update pushed for more reducer :-)

**Comment by charlotte12l** [comment]
Thanks! I totally forgot to modify the file pythran/pythonic/include/numpy/mean.hpp correspondingly...

So smart to use `make_scalar_pshape`, I only knew creating an array and fill it lol.

I added some comments on the following codes- if node.value is int, can we still do str(node.value).lower()? 
```
if node in self.immediates:
            assert isinstance(node.value, int)
            return "std::integral_constant<%s, %s>{}" % (
                PYTYPE_TO_CTYPE_TABLE[type(node.value)], str(node.value).lower())
```

**Comment by serge-sans-paille** [comment]
> I added some comments on the following codes- if node.value is int, can we still do str(node.value).lower()?

yes. `str(<some-int-value>)` will give a number as string, which is unaffected by `lower()`

**Comment by serge-sans-paille** [comment]
Obsoleted by #1869

**Pull request closed without merging**